File: apps/fomc-dashboard/src/lib/playbook.ts
// Pure conditional-stats helpers for the Playbook panel: "given the current stated
// regime, what historically happened around FOMC decisions like the next one?"
// Historical conditional evidence only — not a forecast, not investment advice.

import type { FomcEvent } from "../types";

/** Return windows shown in the playbook grid, in display order. */
export const PLAYBOOK_WINDOWS = ["d0", "p1", "p3", "p5", "p10"] as const;
export type PlaybookWindow = (typeof PLAYBOOK_WINDOWS)[number];

export type PlaybookAsset = "spy" | "tlt";

/** Numeric return column of a FomcEvent, e.g. "spy_p5". */
export type ReturnKey = `${PlaybookAsset}_${PlaybookWindow}`;

export interface ConditionOptions {
  regime: string;
  action?: string;
}

/**
 * Filter events to the conditioning set: same regime (empty/missing regime is
 * treated as "Unknown"), optionally same action, and NEVER emergency meetings —
 * unscheduled decisions are a different animal and would pollute the base rates.
 */
export function conditionEvents(events: FomcEvent[], opts: ConditionOptions): FomcEvent[] {
  const wantRegime = opts.regime || "Unknown";
  return events.filter((e) => {
    if (e.emergency) return false;
    if ((e.regime || "Unknown") !== wantRegime) return false;
    if (opts.action && e.action !== opts.action) return false;
    return true;
  });
}

export interface WindowStats {
  n: number;
  mean: number | null;
  median: number | null;
  /** Share of observations > 0, in [0, 1]. Null when n = 0. */
  winRate: number | null;
}

function round2(v: number): number {
  return Math.round(v * 100) / 100;
}

/** Null-safe summary stats over one return column. */
export function windowStats(events: FomcEvent[], key: ReturnKey): WindowStats {
  const vals: number[] = [];
  for (const e of events) {
    const v = e[key];
    if (typeof v === "number" && !Number.isNaN(v)) vals.push(v);
  }
  const n = vals.length;
  if (n === 0) return { n: 0, mean: null, median: null, winRate: null };

  const mean = round2(vals.reduce((a, b) => a + b, 0) / n);
  const sorted = [...vals].sort((a, b) => a - b);
  const mid = Math.floor(n / 2);
  const median = round2(n % 2 === 1 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2);
  const winRate = vals.filter((v) => v > 0).length / n;
  return { n, mean, median, winRate };
}

export type PlaybookGrid = Record<PlaybookAsset, Record<PlaybookWindow, WindowStats>>;

/** Full SPY/TLT × window stats grid for an (already conditioned) event set. */
export function playbookTable(events: FomcEvent[]): PlaybookGrid {
  const build = (asset: PlaybookAsset): Record<PlaybookWindow, WindowStats> => {
    const row = {} as Record<PlaybookWindow, WindowStats>;
    for (const w of PLAYBOOK_WINDOWS) {
      row[w] = windowStats(events, `${asset}_${w}` as ReturnKey);
    }
    return row;
  };
  return { spy: build("spy"), tlt: build("tlt") };
}

File: packages/fomc-server/src/fomc_server/_playbook.py
"""Pure conditional-stats + event-color helpers (port of the dashboard's lib/playbook.ts
and lib/colorDay.ts logic). "Given the current stated regime, what historically happened
around FOMC decisions like the next one?" Historical conditional evidence only.
"""

from __future__ import annotations

from collections import Counter
from statistics import median as _median
from typing import Any

PLAYBOOK_WINDOWS: tuple[str, ...] = ("d0", "p1", "p3", "p5", "p10")
PLAYBOOK_ASSETS: tuple[str, ...] = ("spy", "tlt")

WINDOW_DEFINITIONS: dict[str, str] = {
    "d0": "decision-day return, % (close vs prior close)",
    "p1": "return over the 1 trading day after the decision, %",
    "p3": "return over the 3 trading days after the decision, %",
    "p5": "return over the 5 trading days after the decision, %",
    "p10": "return over the 10 trading days after the decision, %",
}

Event = dict[str, Any]


def condition_events(events: list[Event], regime: str, action: str | None = None) -> list[Event]:
    """Filter to the conditioning set: same regime (missing regime -> "Unknown"),
    optionally same action, and NEVER emergency meetings — unscheduled decisions
    are a different animal and would pollute the base rates."""
    want = regime or "Unknown"
    out: list[Event] = []
    for e in events:
        if e.get("emergency"):
            continue
        if (e.get("regime") or "Unknown") != want:
            continue
        if action and e.get("action") != action:
            continue
        out.append(e)
    return out


def window_stats(events: list[Event], key: str) -> dict[str, Any]:
    """Null-safe summary stats over one return column (e.g. "spy_p5")."""
    vals = [v for e in events if isinstance(v := e.get(key), (int, float)) and v == v]
    n = len(vals)
    if n == 0:
        return {"n": 0, "mean": None, "median": None, "win_rate": None}
    return {
        "n": n,
        "mean": round(sum(vals) / n, 2),
        "median": round(_median(vals), 2),
        "win_rate": round(sum(1 for v in vals if v > 0) / n, 4),
    }


def playbook_table(events: list[Event]) -> dict[str, dict[str, Any]]:
    """Full SPY/TLT x window stats grid for an (already conditioned) event set."""
    return {
        asset: {w: window_stats(events, f"{asset}_{w}") for w in PLAYBOOK_WINDOWS}
        for asset in PLAYBOOK_ASSETS
    }


def action_counts(events: list[Event]) -> dict[str, int]:
    return dict(Counter(str(e.get("action")) for e in events))


def color_of(spy_d0: float | None, tlt_d0: float | None) -> str | None:
    """SPY/TLT day-0 color. Zero counts as up. Null inputs -> None (no color).

    green: SPY>=0 & TLT>=0 | orange: SPY>=0 & TLT<0 | blue: SPY<0 & TLT>=0 | red: both<0.
    """
    if spy_d0 is None or tlt_d0 is None:
        return None
    if spy_d0 >= 0:
        return "green" if tlt_d0 >= 0 else "orange"
    return "blue" if tlt_d0 >= 0 else "red"
